Fix `WaveformTestBeatmap`s potentially providing disposed track

frenzibyte opened the pull request:
This was noticed by @peppy in [`TestSceneTapTimingControl` in this branch](https://github.com/ppy/osu/blob/b09931327e0562ece51c2120a25cf94ea779218d/osu.Game.Tests/Visual/Editing/TestSceneTapTimingControl.cs) by re-running the test scene at least once, resulting in an `ObjectDisposedException` due to the test scene operating on a disposed track.

Reason why this happens boils down to the track-transferring logic in `MusicController`, in which working beatmaps with identical audio files can have the same track from the last selected beatmap.

In `WaveformTestBeatmap`s, the track is retrieved using a local track store that's disposed on a finaliser. Which means it's not safe to transfer that track to another working beatmap, as the track will be disposed once the previous one is dismissed and finalised by GC.

`ClockBackedTestWorkingBeatmap` should be affected by this as well, but `MusicController` handles that by not allowing to transfer virtual tracks (using the `Track.IsDummyDevice` property).

This PR refactors the track-transferring logic to allow for both `WaveformTestBeatmap` and `ClockBackedTestWorkingBeatmap` to explicitly disallow transferring tracks, and remove the necessity of a `Track.IsDummyDevice` conditional (as it doesn't make sense on face value, until you find the context behind it which is being retrieved by local track stores disposed on finaliser).

frenzibyte committed:
Refactor track transferring logic for ability to override and disallow

frenzibyte committed:
Disallow transferring track from test `WorkingBeatmap`s which have local stores

frenzibyte committed:
Use `==` instead of `??`

frenzibyte commented on osu.Game/Beatmaps/WorkingBeatmap.cs:
```diff
@@ -126,11 +126,19 @@ public void PrepareTrackForPreviewLooping()
-        public void TransferTrack([NotNull] Track track) => this.track = track ?? throw new ArgumentNullException(nameof(track));
+        /// <param name="target">The target working beatmap to transfer this track to.</param>
+        /// <returns>Whether the track is valid and has been transferred to this working beatmap.</returns>
+        public virtual bool TryTransferTrack([NotNull] WorkingBeatmap target)
```
This `TransferTrack` method has been reversed in direction from inwards (transfer track from given one to ours) to outwards (transfer our track to given beatmap). 

This is necessary for `ClockBackedTestWorkingBeatmap` and `WaveformTestBeatmap` to disallow transferring their unsafe track to other working beatmaps.

We could potentially define another `ReceiveTransferredTrack` that's called on the target working beatmap, but I don't think there'll ever be a case where that's necessary.

bdach commented on osu.Game/Beatmaps/WorkingBeatmap.cs:
```diff
@@ -126,11 +126,19 @@ public void PrepareTrackForPreviewLooping()
-        /// <param name="track">The track to transfer.</param>
-        public void TransferTrack([NotNull] Track track) => this.track = track ?? throw new ArgumentNullException(nameof(track));
+        /// <param name="target">The target working beatmap to transfer this track to.</param>
+        /// <returns>Whether the track is valid and has been transferred to this working beatmap.</returns>
```
```suggestion
        /// <returns>Whether the track has been transferred to the <paramref name="target"/>.</returns>
```

Dunno why but the "validity" part of the original xmldoc is throwing me off here. The latter change is fixing a straight-up error (transfer is happening *away* from this instance, not *to* it).

frenzibyte replied:
Whoops, missed to reword this part while I was changing direction of this method.

bdach commented on osu.Game/Overlays/MusicController.cs:
```diff
@@ -290,15 +290,8 @@ private void changeBeatmap(WorkingBeatmap newWorking)
 
             current = newWorking;
 
-            if (!audioEquals || CurrentTrack.IsDummyDevice)
```
Not sure I wholly agree with the `IsDummyDevice` check being removed due to "not making sense at face value". Dummy/virtual tracks are cheap to recreate, and so it makes *some* sense to not transfer virtual tracks to avoid lifetime/shared reference bugs.

frenzibyte replied:
You mention "lifetime/shared reference bugs", but the hope is that we never want to let that happen in the first place unless it makes sense (i.e. `WaveformTestBeatmap`/`ClockBackedTestWorkingBeatmap` storing their track store locally), instead of just adding an early-guard blindly like that and potentially hiding greater issues at hand. Unsure.

bdach replied:
If we want to be as conservative about transferring tracks around as possible, then it makes sense to never transfer tracks which are cheap to construct and don't need the optimisation, no?

frenzibyte replied:
I mean, the reason why we have never seen this issue of potentially transferring disposed tracks on `ClockBackedTestWorkingBeatmap` is because of this specific condition, until the issue became visible with `WaveformTestBeatmap` which returns non-virtual tracks.

But coming at it from a non-testing perspective, I suppose it makes sense that we don't have to transfer virtual tracks given that they are cheap and don't need the "optimisation" this method provides.

frenzibyte committed:
Fix wrong return xmldoc

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

frenzibyte committed:
Bring back virtual track condition given its cheapness

Will still keep the override in `ClockBackedTestWorkingBeatmap` because
it still relies on a local track store and will fail the moment it uses
a non-virtual track.

frenzibyte committed:
Use `Track` to ensure its loaded before transferring

bdach commented on osu.Game/Beatmaps/WorkingBeatmap.cs:
```diff
@@ -126,11 +126,19 @@ public void PrepareTrackForPreviewLooping()
+        /// <returns>Whether the track has been transferred to the <paramref name="target"/>.</returns>
+        public virtual bool TryTransferTrack([NotNull] WorkingBeatmap target)
+        {
+            if (BeatmapInfo?.AudioEquals(target.BeatmapInfo) != true || Track.IsDummyDevice)
```
I see that a17eed64 has sparked a change from `track` to `Track`, to ensure that the track has loaded - but I'm not sure that this is the right way to go here - would it not make sense to just *not* transfer the track if it's not loaded, and let the next `WorkingBeatmap` do it?

frenzibyte replied:
The way I'm seeing it here, and what `MusicController` looks to have been doing prior to this PR, is that if we are requested for transferring a track, we can't possibly be in a non-loaded state, and we should definitely throw instead of not doing anything.

bdach replied:
So it was just a safety change on your end? It wasn't sparked by any actual issues after the other changes?

frenzibyte replied:
Yeah, noticed it while looking at the other suggestions.

peppy reviewed (approved):
I think this is fine
